### PR TITLE
Threaded Motion Controllers

### PR DESCRIPTION
**Pull request by carturn**
Basically, when we call setPosition on a PostionEncodedMotor, or tell the chassis to rotate to a position, we want it to stay there without a dedicated command calling that function periodically.
This pull request would accomplish this by having motion controllers do their calculation in a separate thread with a constant update frequency (I think 20Hz is optimal, as that is basically the same as the rest of the robot's systems) and push that information to motors.

### CODE REVIEW

**Comment by c0d3rman** [comment]
Please drop first cycle to eliminate twitches. (Or constantly calculate D when not enabled)

File: custom/motioncontrollers/MotionController.java [hunk 1]
```diff
@@ -63,14 +89,33 @@ public MotionController(PIDSource source) {
 	 *
 	 * @warning this does not indicate sensor errors
 	 */
-	public abstract void reset();
+	public final void reset() {
+		resetMC();
+		setpoint = sensor.pidGet();
+		synchronized (justReset) {
+			justReset = true;
+		}
```
Comment (leijurv) [inline custom/motioncontrollers/MotionController.java]: Useless for two reasons: synchronizing on a field that changes doesn't do anything, and setting the value of something doesn't need to be synchronized (what matters for synchronizing is blocks of code that require the value to remain the same)

File: custom/motioncontrollers/MotionController.java [hunk 2]
```diff
@@ -217,4 +282,41 @@ public void disable() {
...
+	 */
+	protected class MotionControllerTask extends TimerTask {
+		@Override
+		public void run() {
+			try {
+				double value = getSafely(); // Always calculate MC output
+				if (justReset) {
+					synchronized (justReset) {
+						justReset = false;
+					}
```
Comment (leijurv) [inline custom/motioncontrollers/MotionController.java]: Checking the value outside of a synchronized context, then SETTING the value in its own synchronized context????????

File: custom/motioncontrollers/MotionController.java [hunk 3]
```diff
@@ -34,6 +42,9 @@
 	 */
 	public MotionController(PIDSensor sensor) {
 		this.sensor = sensor;
+		output = null;
+		timer = new Timer();
+		task = new MotionControllerTask();
```
Comment (ave4224) [inline custom/motioncontrollers/MotionController.java]: The reason to use a ScheduledThreadPoolExecutor is to prevent concurrent tasks from delaying each other because Timers use a single thread. However, MotionControllers only have one task each so as long as each has its own instance of a Timer each loop will have its own thread.

File: Util.java [hunk 4]
```diff
@@ -1,5 +1,6 @@
 package org.usfirst.frc4904.standard;
 
+
```
Comment (ajnadel) [inline Util.java]: Somebody check this against the formatter.

**Comment by ave4224** [comment]
Ready to Merge. I tested it and it is able to set the wheel to a certain angle. I did not well tune PID but rather did a proof of concept.

File: Util.java (hunk 4 above)
Comment (leijurv) [inline Util.java]: yes that's valid when i had my nested ternary pull request it adde the same newline

File: commands/motor/MotorPositionSet.java [hunk 5]
```diff
@@ -69,6 +67,9 @@ protected void execute() {
 	}
 	
 	@Override
+	protected void execute() {}
```
Comment (leijurv) [inline commands/motor/MotorPositionSet.java]: no

File: Util.java (hunk 4 above)
Comment (c0d3rman) [inline Util.java]: Can confirm

**Comment by c0d3rman** [comment]
This is a very large change, and I'd like to get as many people to sign off on it as possible. @leijurv, @ave4224, @ajnadel, @carturn, @widdakay and anyone else, please reread critically - consider edge cases, think about better structure for future maintenance/graceful failure, add as much documentation as possible, etc. Once this has your golden stamp of approval, please approve in a review.

File: commands/motor/MotorPositionSet.java (hunk 5 above)
Comment (carterturn) [inline commands/motor/MotorPositionSet.java]: Yes.
TL;DR: It would be problematic for the motion controller to silently fail; the fallbackCommand should be started on sensor failure, regardless of whether it was during initialization or while running.
Given that the way some errors are detected is through timeouts, it is actually more likely that an error will become apparent while running than during initialization. While the motion controller will disable itself either way, it does not have any ability to trigger the fallback without the command actively checking for errors. It might be possible to implement that, but I am inclined to prefer to separate the fallback command from the controller.

**Comment by carterturn** [comment]
I think it would be nice to make the ChassisControllers be able to use the setOutput of the motion controller, as this would allow the movement to be slightly more precise.
Also, I think we need to either ensure there is a standard for motion controllers being enabled and disabled or add the relevant enable and disable calls to all commands. At the moment, there are a few commands that either enable or disable a motion controller. They do not appear to restore to the original state after running, which is suboptimal. I think it would be safer to have the initialize in all core motor and chassis commands set the motion controller to enabled or disabled, especially as with setOutput the ending of a command need not necessarily trigger the ending of a motion control.

**Comment by ave4224** [comment]
It might be useful to include [this](http://pastebin.com/DbHFGmkL), but I don't know how I would make a test case.

**Comment by carterturn** [comment]
@ave4224 I think it would be better to just call getSafely on the motion controller. There is not really an advantage to having the motion controller write to a class, then reading a variable from that class.

**Comment by ave4224** [comment]
That makes sense.
On Jan 23, 2017, at 6:42 PM, Carter Turnbaugh <notifications@github.com<mailto:notifications@github.com>> wrote:


@ave4224<https://github.com/ave4224> I think it would be better to just call getSafely on the motion controller. There is not really an advantage to having the motion controller write to a class, then reading a variable from that class.

File: custom/MCChassisController.java [hunk 6]
```diff
@@ -8,15 +8,15 @@
...
+public class MCChassisController implements ChassisController {
 	protected ChassisController controller;
 	protected double maxDegreesPerSecond;
 	protected double targetYaw;
 	protected double lastUpdate;
 	protected IMU imu;
 	protected MotionController motionController;
 	
-	public PIDChassisController(ChassisController controller, IMU imu, MotionController motionController, double maxDegreesPerSecond) {
+	public MCChassisController(ChassisController controller, IMU imu, MotionController motionController, double maxDegreesPerSecond) {
```
Comment (ajnadel) [inline custom/MCChassisController.java]: The official pronunciation of this class is "mick chassis controller" as if it were a McDonalds product.

**Comment by ajnadel** [comment]
Tested!
